### PR TITLE
Add Datalake to FlowEHR

### LINKED ISSUE #160: Data Lake to store ingested and transformed data
As a data engineer I want a landing zone for raw data, and a place to store transformed data.

This should follow the medallion approach as outlined here -https://learn.microsoft.com/en-us/azure/databricks/lakehouse/medallion and here https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/scenarios/cloud-scale-analytics/best-practices/data-lake-zones

As each deployment requirements may differ, the zones and containers created at the time of deployment should be configurable.

For example:

| Zone | Container
--|--
| Bronze | Raw
| Bronze | Conformed
| Silver | Validated 
| Silver | Pseudonymised
| Gold | Research Ready
| Gold | Internal Analytics

The data lake should be accessible from Databricks cluster (mount point) and ADF (linked service),

I have started some exploration here:
https://github.com/UCLH-Foundry/FlowEHR/compare/main...marrobi:FlowEHR:marrobi/adls-testing 

@nels @tanya-borisova @damoodamoo @anastasiakuzn thoughts? I might look to do a PR based of the exploration work.

### LINKED ISSUE #308: Pre-commit terragrunt validate / lint is broken with missing provider errors
**Describe the bug**
In a new fork, the listing CI checks break with the following errors from pre-commit:

```
TFLint in apps/app/aad_app/:
4 issue(s) found:

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on  line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_version.md

Warning: Missing version constraint for provider "random" in "required_providers" (terraform_required_providers)

  on main.tf line 15:
  15: resource "random_uuid" "webapp_oauth2_id" {}

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "azuread" in "required_providers" (terraform_required_providers)

  on main.tf line 59:
  59: resource "azuread_application_password" "webapp_auth" {

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.2.2/docs/rules/terraform_required_providers.md

Warning: Missing version constraint for provider "azurerm" in "required_providers" (terraform_required_providers)

  on main.tf line 70:
  70: resource "azurerm_key_vault_secret" "client_secret" {
```

and

```
Terraform validate.......................................................Failed
- hook id: terraform_validate
- exit code: 1

'terraform init' failed, 'terraform validate' skipped: apps/app
Initializing modules...
- aad_app in aad_app

Initializing provider plugins...
- Reusing previous version of hashicorp/azuread from the dependency lock file
- Reusing previous version of hashicorp/random from the dependency lock file
- Reusing previous version of hashicorp/azurerm from the dependency lock file
- Finding integrations/github versions matching "5.18.3"...
- Finding hashicorp/external versions matching "2.3.1"...
- Installing hashicorp/external v2.3.1...
- Installed hashicorp/external v2.3.1 (signed by HashiCorp)
- Installing integrations/github v5.18.3...
- Installed integrations/github v5.18.3 (signed by a HashiCorp partner, key ID 38027F80D7FD5FB2)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html
╷
│ Error: Failed to query available provider packages
│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/azuread: locked provider registry.terraform.io/hashicorp/azuread
│ 2.39.0 does not match configured version constraint 2.35.0; must use
│ terraform init -upgrade to allow selection of new versions
╵

╷
│ Error: Failed to query available provider packages
│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/random: locked provider registry.terraform.io/hashicorp/random
│ 3.5.1 does not match configured version constraint 3.4.3; must use
│ terraform init -upgrade to allow selection of new versions
╵

╷
│ Error: Failed to query available provider packages
│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/azurerm: locked provider registry.terraform.io/hashicorp/azurerm
│ 3.63.0 does not match configured version constraint 3.47.0; must use
│ terraform init -upgrade to allow selection of new versions
╵
```

**Expected behavior**
Pre-commit and linting CI checks should pass

**Additional context**
I believe this is likely related to the introduction of the `aad_app` submodule. Due to an issue with the `app` child module not selecting the correct GitHub terraform provider, a `terraform.tf` file is generated for the submodule by terragrunt. It looks like Terraform now expects the same for the child within that, and is failing as it can't find any provider specifications (as terragrunt isn't currently generating this.)

### PR DESCRIPTION
**Pull request by jjgriff93**
- Resolves #160 
- Fixes #308

## What is being addressed

This PR builds upon https://github.com/UCLH-Foundry/FlowEHR/pull/198 and adds an optional datalake deployment to FlowEHR, which can be enabled and configured via the FlowEHR `config.yaml`:

```
transform:
  datalake:
    zones:
      - Bronze
      - Silver
      - Gold
```

This adds an out of the box data lake storage set-up to be used with Databricks pipelines as part of enabling a [Data Lakehouse](https://www.databricks.com/glossary/data-lakehouse) pattern.

## How is this addressed

- Add new Datalake module inc. ADLS Gen2 storage
- Databricks MSI & spark config to authenticate to ADLS
- Data factory linked service for ADLS via MSI
- Private endpoint and vnet link

## Has the CIS document been consulted / updated?

- [X] Added Storage Account (ADLS) to Data Transformation section

### CODE REVIEW

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5644335681 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1647741391) from @jjgriff93)

File: infrastructure/transform/datalake/adls.tf [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+# Add IP exception when deploying locally
+resource "azurerm_storage_account_network_rules" "adls" {
+  storage_account_id = azurerm_storage_account.adls.id
+  default_action     = "Deny"
+  ip_rules           = var.tf_in_automation == true ? [] : [var.deployer_ip]
+}
+
+# Create filesystem for each zone
```
Comment (tanya-borisova) [inline infrastructure/transform/datalake/adls.tf:58]: Nit: s/filesystem/container

**Review by tanya-borisova — APPROVED** [review]
Looks great! One tiny nit

File: infrastructure/transform/datalake/adls.tf (hunk 1 above)
Comment (jjgriff93) [reply infrastructure/transform/datalake/adls.tf]: Clarified

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5647969431 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1648321426) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5648984244 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1648485562) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5666779844 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1651282360) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5681372654 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1653664463) from @tanya-borisova)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5682404835 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1653836953) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5683032494 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1653948464) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5692859327 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1655749862) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5693100064 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1655791511) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5693219621 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1655812080) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5693893214 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1655933730) from @jjgriff93)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5694734709 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1656088917) from @tanya-borisova)

**Comment by jjgriff93** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/5694807943 (with refid `0ef1bf82`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/310#issuecomment-1656099788) from @tanya-borisova)

**Pull request merged**